service_configuration_lib/utils.py

```diff
@@ -85,3 +113,46 @@ def ephemeral_port_reserve_range(preferred_port_start: int, preferred_port_end:
 
 def get_random_string(length: int) -> str:
     return ''.join(random.choice(string.ascii_lowercase + string.digits) for _ in range(length))
+
+
+def generate_pod_template_path():
+    return POD_TEMPLATE_PATH.format(file_uuid=uuid.uuid4().hex)
+
+
+def create_pod_template(pod_template_path, app_base_name):
+    document = POD_TEMPLATE.format(
+        spark_pod_label=get_k8s_resource_name_limit_size_with_hash(f"exec-{app_base_name}"),
+    )
+    parsed_pod_template = yaml.safe_load(document)
+    with open(pod_template_path, "w") as f:
+        yaml.dump(parsed_pod_template, f)
+
+
+def get_k8s_resource_name_limit_size_with_hash(name: str, limit: int = 63, suffix: int = 4) -> str:
+    """ Returns `name` unchanged if it's length does not exceed the `limit`.
+        Otherwise, returns truncated `name` with its hash of size `suffix`
+        appended.
+
+        base32 encoding is chosen as it satisfies the common requirement in
+        various k8s names to be alphanumeric.
+
+        NOTE: This function is the same as paasta/paasta_tools/kubernetes_tools.py
+    """
+    if len(name) > limit:
+        digest = hashlib.md5(name.encode()).digest()
+        hashed = base64.b32encode(digest).decode().replace('=', '').lower()
+        return f'{name[:(limit-suffix-1)]}-{hashed[:suffix]}'
+    else:
+        return name
+
```

Comment:
Are the functions above moved to here because they will be used in multiple different repos? I think all the current 3 run types `paasta spark-run`, `tron_tools`, and `spark_tools` should call `get_spark_conf()` in service-configuration-lib to make things consistent except logic specific to each run types. IIUC, both `pod template` and `k8s resource name limitation` for executor pods applies to all run types.

Reply:
This logic was in paasta spark-run and didn't exist in service configuration lib. I have moved it here because this is common and we are calling `get_spark_conf` everywhere

Reply:
++ I guess the idea is not to import and call functions in `service_configuration_lib.utils` by other repos, i.e. these will be only used by this repo itself

Reply:
That's correct